IPIP/0000-cidv2-tagged-content-identifiers.md

```diff
@@ -0,0 +1,130 @@
+Create a new [CID](https://github.com/multiformats/cid) version (CIDv2, informally "Tagged Content Identifiers") which combines a data Multicodec-Multihash pair (the pointer) and a metadata Multicodec-Multihash pair (the tag) to create content-addresses with expressive contexts.
+
+## Motivation
+
+Currently, CIDv1 data is described by a multicodec content type. However, this is meant to describe the overall format of the serialized data e.g. the `dag-cbor` IPLD encoding, and not more specific information such as a data schema or type. For example, it can be useful to have raw IPLD data contextualized by its [IPLD schema](https://ipld.io/docs/schemas/intro/). Since multicodecs are limited to 9 bytes  by the [unsigned-varint spec](https://github.com/multiformats/unsigned-varint#practical-maximum-of-9-bytes-for-security), the available codec space is generally too small to encode such metadata.
+
+## Detailed design
+
+Our solution is a new CID version which contains two multicodec-multihash pairs, one pair for data and another for metadata. The metadata multicodec would be able to concisely describe a space of metadata tags where the specific tag would then be further specified by the multihash. This could be implemented as follows in Rust:
+
+```rust
+pub struct CidV2<const S: usize, const M: usize> {
+    /// The data multicodec
+    data_codec: u64,
+    /// The data multihash
+    data_hash: Multihash<S>,
+    /// The metadata codec
+    meta_codec: u64,
+    /// The metadata multihash
+    meta_hash: Multihash<M>
+}
+```
+
+It would serialize as follows:
+
+```
+<cidv2> ::= <multicodec-cidv2><multicodec-data-content-type><multihash-data><multicodec-metadata-content-type><multihash-metadata>
+```
+
+with a multibase prefix when represented in text.
+
+For example, suppose you want a CID which points to a piece of IPLD data and its [IPLD schema](https://ipld.io/docs/schemas/). Let's say you have the schema `Trit`, with a particular integer representation 
+
+```
+type Trit union {
+  | True ("1")
+  | False ("2")
+  | Unknown ("0")
+} representation int
+```
+
+which corresponds to the Ipld data: `Ipld::Num(1)`, `Ipld::Num(2)`, `Ipld::Num(0)`.
+
+While you could in principle propose a new multicodec for `Trit`, this might be not suitable if `Trit` is a temporary or ephemeral structure, or if you have a large number of different schemas (For instance, in Lurk-lang's content-addressing we would need to reserve 16-bits of the multicodec table, or 2^16 distinct multicodecs).
+
+However, since IPLD schemas can be [represented as JSON](https://ipld.io/specs/schemas/#dsl-vs-dmt) and hashed, with a CIDv2 we could reserve a single IPLD schema multicodec, along with the codec for the data representation (such as dag-cbor)
+We could then use the above CIDv2 definiton to create a pointer to any Schema+Data pair:
+
+```
+CidV2 { 
+  data_codec: 0x71,
+  data_hash: <data_multihash>,
+  meta_codec : 0x3e7ada7a, 
+  meta_hash: <schema-multihash> 
```

Comment:
> This looks great, but what happens if my schema starts becoming large?

It doesn't sound like this is a CIDv2 specific issue. I can store a 3MiB dag-cbor IPLD object on IPFS and generate a CIDv1 with its sha256 multihash, right? In terms of transport, I think CIDv2 will just behave like a pair of CIDv1s

Reply:
> It doesn't sound like this is a CIDv2 specific issue. 

No, it is a specific issue with this CIDv2 proposal. Nowhere else do we use codec identifiers as data types, we use them as deserialization types. As a result there is no notion of a data type changing or becoming too big, that becomes an application layer concern. For example, an object can represent a UnixFS directory whether it is a single directory block or the root of a sharded HAMT.

By using the code as a nominative type rather than a description of how to deserialize the data you've navigated into a position where there's nowhere to identify both the type of the data and how to get it as a multiblock data structure without another level of indirection. However, that level of indirection could similarly be used instead of CIDv2 entirely (see alternative proposal).

Reply:
> Nowhere else do we use codec identifiers as data types, we use them as deserialization types.

Using a codec identifier as a datatype isn't an essential (or actually even intended) part of this proposal, so I'm absolutely happy to make any changes you suggest to better align with how multicodecs should be used. 

For example, replacing the `0x3e7ada7a` example codec for `ipld-schema` with `dag-cbor`, and in general specifying that the metadata codec should refer to the metadata format would be totally fine for the intended use-cases